cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -25,6 +25,52 @@
 
 CrossEntropyPair = NamedTuple('CrossEntropyPair', [('num_cycle', int),
                                                    ('xeb_fidelity', float)])
+SpecklePurityPair = NamedTuple('SpecklePurityPair', [('num_cycle', int),
+                                                     ('purity', float)])
+
+
+def purity_from_probabilities(
+        hilbert_space_dimension: int,
+        probabilities: Sequence,
+) -> float:
+    """Purity estimator from speckle purity benchmarking.
+
+    Estimates purity from empirical probabilities of observed bitstrings.
+
+    This estimator assumes that the circuit used in experiment is sufficiently
+    scrambling that its output probabilities follow the Porter-Thomas
+    distribution. This assumption holds for typical instances of random quantum
+    circuits of sufficient depth.
+
+    The state resulting from the experimental implementation of the circuit
+    is modeled as
+
+        ρ = p |𝜓⟩⟨𝜓|  + (1 - p) I / D
+
+    where |𝜓⟩ is a pure state, I / D is the maximally mixed state, and p is
+    between 0 and 1. The purity of this state is given by p**2. If p = 1, then
+    the bitstring probabilities are modeled as being drawn from the
+    Porter-Thomas distribution, with probability density function given by
+
+        f(x) = (D - 1) (1 - x)**(D - 2).
+
+    The mean of this distribution is 1 / D and its variance is
+    (D - 1) / [D**2 (D + 1)]. In general, the variance of the distribution
+    is multipled by p**2. Therefore, the purity can be computed by dividing
+    the variance of the empirical probabilities by the Porter-Thomas
+    variance (D - 1) / [D**2 (D + 1)].
+
+    Args:
+        hilbert_space_dimension: Dimension of the Hilbert space on which the
+            quantum circuits acts.
+        probabilities: Empirical probabilities of bitstrings observed in
+            experiment.
+    Returns:
+        Estimate of the purity of the state resulting from the experimental
+        implementation of a quantum circuit.
+    """
+    D = hilbert_space_dimension
+    return np.var(probabilities) * D**2 * (D + 1) / (D - 1)
```

Comment:
Optional: I assume you're creating the local variable just to shorten the return statement. Here's an alternative that has an additional benefit of explaining to the reader what's happening:

```python
pt_var = (hilbert_space_dimension - 1) / (hilbert_space_dimension + 1) / hilbert_space_dimension**2
return np.var(probabilities) / pt_var
```

Reply:
Even with your suggestion we would have to split the formula for `pt_var` over multiple lines, which I think is much less readable. I've taken your suggestion to create a `pt_var` variable, but kept the variable `D`.